### ISSUE TITLE
'skopeo copy' gets stuck uploading images to a Docker registry on Fedora >= 40 hosts

### ISSUE THREAD

**Issue by debarshiray** [containers/podman#22575]
This was [observed](https://github.com/containers/toolbox/pull/1468) when trying to add Fedora 40 to the list of host operating systems on which the upstream Toolbx CI is run.  Eventually, it started happening on Fedora Rawhide hosts too.

The Toolbx test suite sets up a local temporary Docker registry to test some behaviour of the `toolbox(1)` binary.

First, it sets up a TLS key pair:
```
$ mkdir /var/tmp/foo
$ openssl req \
    -newkey rsa:4096 \
    -nodes -sha256 \
    -keyout "/var/tmp/foo/domain.key" \
    -addext "subjectAltName= DNS:localhost" \
    -x509 \
    -days 365 \
    -subj "/" \
    -out "/var/tmp/foo/domain.crt"
...
$ mkdir --parents ~/.config/containers/certs.d/localhost\:50000
$ cp /var/tmp/foo/domain.crt ~/.config/containers/certs.d/localhost\:50000
```

Second, it runs a Docker registry in a container:
```
$ podman run \
    --detach \
    --rm \
    --name docker-registry \
    --privileged \
    --volume /var/tmp/foo:/certs \
    -e REGISTRY_HTTP_ADDR=0.0.0.0:443 \
    -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
    -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
    -p 50000:443 \
    docker.io/library/registry:2
...
```

Finally, it populates the Docker registry with an image:
```
$ podman pull registry.fedoraproject.org/fedora:34
...
$ skopeo copy containers-storage:registry.fedoraproject.org/fedora:34 docker://localhost:50000/fedora:34
Getting image source signatures
Copying blob 883e787c00d4 done   | 
<stuck>
```

That last step to upload an older Fedora image to populate the local temporary Docker registry always gets stuck on hosts running Fedora 40 and newer, but doesn't on hosts running older Fedora, such as 38 and 39.

For various odd requirements and quirks, the Toolbx test suite specifically uploads an older Fedora based image to simplify the implementation.  That said, I reproduced it also with the `fedora:35`, `fedora:36`, `fedora:37`, `fedora:38` and `fedora:39` images on a Fedora 40 host.  The `fedora:40` image didn't get stuck on a Fedora 40 host.

All my tests were done using Skopeo 1.15.0 on Fedora 41, 40, 39 and 38 hosts.  The corresponding RPMs were:
```
skopeo-1.15.0-1.fc41.x86_64
skopeo-1.15.0-1.fc40.x86_64
skopeo-1.15.0-1.fc39.x86_64
skopeo-1.15.0-1.fc38.x86_64
```

**Comment by debarshiray** [comment]
> Did you try to push with podman push, does this hang as well?

No, I didn't.  I will try it now.

> Is the https setup required or does it also hang for unencrypted traffic?

I don't think HTTPS is necessary, but I haven't actually tried plain text traffic.

> Juts looking to make the reproducer a bit simpler.

This part of the test suite wasn't written by me.  So I am also trying to remove the unnecessary bits as I go along.  eg., the original code created a Docker registry that requires authentication.

I will try to get rid of the HTTPS part.

> In any case you should run podman with something like `--network pasta:--pcap,/tmp/pasta.pcap` which will record all packages send then you can just upload the file here and the pasta folks can have a look based on that.

Okay!

**Comment by sbrivio-rh** [comment]
Could this have something to do with the `host.containers.internal` [issue](https://github.com/containers/podman/issues/19213#issuecomment-2088022381), and the container here would need an additional address?

I'm not exactly sure how skopeo uses pasta here, and how local ports are remapped (is it something going into pasta's configuration as well?).

**Comment by Luap99** [comment]
> Could this have something to do with the host.containers.internal https://github.com/containers/podman/issues/19213#issuecomment-2088022381, and the container here would need an additional address?

I don't think so

Skopeo doesn't use pasta, the setup contains a container registry run inside podman with a forwarded tcp port into the container. The container registry is really just a normal HTTP(S) server with a special API. skopeo just connects to it via the forwarded port and uploads the image data which seems to hang at the end of the upload.

**Comment by mtrmac** [comment]
Also, the TCP connection worked well enough to set up HTTPS, i.e. the packet flow was bidirectional and working, at least for a while. That should rule out at least the long-term-persistent aspects of IP addressing.

**Comment by sbrivio-rh** [comment]
Ah, okay, thanks for clarifying. Let's wait for captures then.

**Comment by debarshiray** [comment]
With plain text HTTP, the reproducer becomes:

Run the Docker registry container with:
```
$ podman run \
      --detach \
      --name docker-registry \
      --publish 5000:5000 \
      --rm \
      docker.io/library/registry:2
...
```

Then try to upload an image to it:
```
$ skopeo copy \
      --dest-tls-verify=false \
      containers-storage:registry.fedoraproject.org/fedora:34 \
      docker://localhost:5000/fedora:34
Getting image source signatures
Copying blob 883e787c00d4 done   | 
<stuck>
```

Note the extra `--dest-tls-verify=false` option.

**Comment by debarshiray** [comment]
> you should run podman with something like `--network pasta:--pcap,/tmp/pasta.pcap` which will record all packages send then you can just upload the file here and the pasta folks can have a look based on that

Here's the file: https://rishi.fedorapeople.org/podman-22575-pasta.pcap

**Comment by sbrivio-rh** [comment]
> Here's the file: https://rishi.fedorapeople.org/podman-22575-pasta.pcap

Weird, there are some packets there, but absolutely no TCP connections or data. It looks like it comes from an idle container or suchlike...?

**Comment by sbrivio-rh** [comment]
By the way, I tried the reproducer, but I'm not familiar with skopeo at all, and on the second command I get:

```
$ skopeo copy       --dest-tls-verify=false       containers-storage:registry.fedoraproject.org/fedora:39       docker://localhost:5000/fedora:39
FATA[0000] initializing source containers-storage:[overlay@/home/sbrivio/.local/share/containers/storage+/run/user/1000/containers]registry.fedoraproject.org/fedora:39: reference "[overlay@/home/sbrivio/.local/share/containers/storage+/run/user/1000/containers]registry.fedoraproject.org/fedora:39" does not resolve to an image ID: identifier is not an image 
```

**Comment by Luap99** [comment]
you have to pull the image first if it does not exists locally `podman pull registry.fedoraproject.org/fedora:39`

**Comment by sbrivio-rh** [comment]
Works for me :/

```
$ skopeo copy       --dest-tls-verify=false       containers-storage:registry.fedoraproject.org/fedora:39       docker://localhost:5000/fedora:39
Getting image source signatures
Copying blob 404a04c0d585 done   | 
Copying config f4f05d0fcc done   | 
Writing manifest to image destination
```

...with Podman and pasta at current `HEAD`, skopeo a bit older (1.13.3). I guess I should try matching the versions used here for all the components.

**Comment by debarshiray** [comment]
I have been using these for my reproducer:
```
$ rpm -q passt podman skopeo
passt-0^20240426.gd03c4e2-1.fc40.x86_64
podman-5.0.2-1.fc40.x86_64
skopeo-1.15.0-1.fc40.x86_64
```

There's definitely an element of non-determinism here.  Skopeo didn't always get stuck on my Fedora 40 Workstation virtual machine, but it often did, while on the [Zuul CI](https://github.com/containers/toolbox/pull/1468) instance it seemed to get stuck almost always.

**Comment by debarshiray** [comment]
I have been doing some printf() debugging within the Skopeo sources to understand the flow, because Skopeo seemed like an easier starting point.

In `vendor/github.com/containers/image/v5/copy/single.go, copyLayers()`, one instance of a `copyLayerHelper()` go subrouting is launched.  This invokes `copyLayer()`, where `diffIDIsNeeded` evaluates to `false`, and we end up in the final fallback copy layer logic.

Here, `vendor/github.com/containers/image/v5/storage/storage_src.go, GetBlob()` gets called, but not `copyLayerFromStream()`.

**Comment by mtrmac** [comment]
That’s different from the earlier traceback, then. There, goroutine 51 is inside `copyBlobFromStream`, well after `GetBlob` has returned.

**Comment by sbrivio-rh** [comment]
> That’s different from the earlier traceback, then. There, goroutine 51 is inside `copyBlobFromStream`, well after `GetBlob` has returned.

By the way, two notable differences between slirp4netns and pasta that might cause different behaviours here are:
- pasta supports window scaling, unlike slirp4netns. Due to this, with slirp4netns, the maximum TCP receive window is limited to 64 KiB (that is, the sender can send at most 64 KiB at a time before waiting from an acknowledgement from the receiver). With pasta, for local connections on a reasonably fast system, you'll see several megabytes of receive windows shortly after a connection is established
- pasta happily works with 64 KiB MTU (and a bit under that for TCP MSS), while slirp4netns should somewhat limit the throughput with that value (which is anyway the default, also for slirp4netns).

All in all, typically, you'll get larger batches of data with pasta, and faster.

**Comment by dgibson** [comment]
> > Here's the file: https://rishi.fedorapeople.org/podman-22575-pasta.pcap
> 
> Weird, there are some packets there, but absolutely no TCP connections or data. It looks like it comes from an idle container or suchlike...?

Actually, that makes sense.  From the original reproducer:

```
$ skopeo copy containers-storage:registry.fedoraproject.org/fedora:34 docker://localhost:50000/fedora:34
```

So we're connecting to the container via a localhost connection, which means it will go via pasta's "splice" path.  That bypasses the normal tap interface, so won't show up in the packet trace.  This also means that from the container's point of view the connection will also come from localhost, which might not be what the registry server as containerized expects, so that's one possible cause of the issue.  Otherwise, I'd suggest this info as next steps:

- `ip a` output from within the registry container (just to make sure there aren't somehow some additional network interfaces confusing the picture.
- A packet dump from the container's `lo` interface.  That might be a bit fiddly to obtain if the container image doesn't include `tcpdump`, or `dumpcap`.  It should be possible by using `nsenter` to enter just the container's netns and do that, though.  There might be an easier podman specific way of doing so, @Luap99 ?
- Checking the exact command line of `pasta` with ps would be helpful.  Do this after creating the container, but before running the stalling command, because...
- ...it's worth checking if the pasta still exists after attempting the reproducer - it's possible pasta is simply crashing.

**Comment by Luap99** [comment]
> A packet dump from the container's lo interface. That might be a bit fiddly to obtain if the container image doesn't include tcpdump, or dumpcap. It should be possible by using nsenter to enter just the container's netns and do that, though. There might be an easier podman specific way of doing so, @Luap99 ?

I really need to start writing some scripts around this given how common this is.
The command would be somthing like `podman unshare nsenter --net=$(podman inspect --format {{.NetworkSettings.SandboxKey}} <container id/name>) <command inside netns...>`
So if we only care about lo traffic then `tcpdump -i lo -w lo.pcap` as command in the container should do it.

> Checking the exact command line of pasta with ps would be helpful. Do this after creating the container, but before running the stalling command, because...

starting podman with `--log-level debug` will print the pasta arguments

> ...it's worth checking if the pasta still exists after attempting the reproducer - it's possible pasta is simply crashing.

Just logically speaking if it crashes the sockets will be closed and skopeo shouldn't hang and instead fail with something like `connection reset`

**Comment by debarshiray** [comment]
There's clearly a race.  I tried inserting a `fmt.Printf()` each in the deferred function calls in `GetBlob()` and one after `GetBlob()` had returned, and suddenly `skopeo copy` wasn't stuck anymore.  Instead it, it ended with an exit code of 1 with this `FATAL` message:
```
FATA[0000] reading blob sha256:883e787c00d4208d75fc3e85d100ce64b517e49a2468f0e7f084cf05d16f3e46: read /var/tmp/container_images_2486861299: file already closed 
```

**Comment by mtrmac** [comment]
What’s the change? I’m rather skeptical.
1. I can’t recall seeing a failure like that. It would be a curious coincidence if you happened to hit one, after years of Skopeo existence, within 2 days or so, _and_ while changing the code.
2. By the time this stream gets down to `copyBlobFromStream` (where that error is almost certainly reported, in `blobPipelineDetectCompressionStep`), the stream is an `io.Reader` without a `Close` method even easily accessible. Now, granted, Go is dynamic enough that something _could_ cast the stream to an `io.Closer` and call `Close`, but why would it?
3. If this causes an error reading the input, how does it explain a hang?

It’s far too attractive for me to suspect that “adding a `Printf` to a deferred function call” happened to turn one of the deferred calls into a non-deferred one.

**Comment by debarshiray** [comment]
> What’s the change? I’m rather skeptical.

Here's a scratch build with the changes I am playing with:
https://koji.fedoraproject.org/koji/taskinfo?taskID=117212076

You can grab the source RPM and look at the changes I made.  Note that I commented out the `fmt.Printf()` invocations in the deferred calls and it still gives me the `FATAL` above.

Leaving aside my nosing around, there were occasions when an unmodified `skopeo(1)` still worked, and didn't get stuck, with the Docker registry container running with `pasta(1)`.  It even made me believe for a while that only copying some images triggered the bug.

**Comment by debarshiray** [comment]
> > > Here's the file: https://rishi.fedorapeople.org/podman-22575-pasta.pcap
> > 
> > 
> > Weird, there are some packets there, but absolutely no TCP connections or data. It looks like it comes from an idle container or suchlike...?
> 
> Actually, that makes sense. From the original reproducer:
> 
> ```
> $ skopeo copy containers-storage:registry.fedoraproject.org/fedora:34 docker://localhost:50000/fedora:34
> ```
> 
> So we're connecting to the container via a localhost connection, which means it will go via pasta's "splice" path. That bypasses the normal tap interface, so won't show up in the packet trace. This also means that from the container's point of view the connection will also come from localhost,

Thanks for that explanation.

> which might not be what the registry server as containerized expects, so that's one possible cause of the issue. Otherwise, I'd suggest this info as next steps:

The [documentation](https://distribution.github.io/distribution/about/deploying/) for the Docker registry does talk about running it locally inside a container.  Although, they talk about Docker containers, not Podman containers.  So, maybe the difference due to `pasta(1)` matters?

>     * `ip a` output from within the registry container (just to make sure there aren't somehow some additional network interfaces confusing the picture.

After reproducing a stuck `skopeo copy`, and with the Docker registry container running:
```
$ podman exec docker-registry ip address
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp1s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65520 qdisc fq_codel state UNKNOWN qlen 1000
    link/ether ae:98:25:49:26:ea brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global noprefixroute enp1s0
       valid_lft forever preferred_lft forever
    inet6 fe80::ac98:25ff:fe49:26ea/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by mtrmac** [comment]
Thanks. That does all look reasonable AFAICT.

I can see one bug in there already, although that one should not make a difference. And the `file already closed` behavior is consistent with that RPM… except I can’t explain it. Looking.

**Comment by debarshiray** [comment]
> Thanks. That does all look reasonable AFAICT.
> 
> I can see one bug in there already, although that one should not make a difference. And the `file already closed` behavior is consistent with that RPM… except I can’t explain it. Looking.

It gets stuck again if I move around the `fmt.Printf()` invocations a bit.  See this scratch build: https://koji.fedoraproject.org/koji/taskinfo?taskID=117213243

Except, this time it's stuck inside `copyLayerFromStream()` within `blobPipelineDecryptionStep()`.  This is puzzling because I was certain that `copyLayerFromStream()` wasn't being called before.

**Comment by debarshiray** [comment]
> Checking the exact command line of pasta with ps would be helpful. Do this after creating the container, but before running the stalling command, because...
> ...it's worth checking if the pasta still exists after attempting the reproducer - it's possible pasta is simply crashing.

The `pasta(1)` process doesn't crash and runs as:
```
/usr/bin/pasta --config-net -t 5000-5000:5000-5000 --dns-forward 169.254.0.1 -u none -T none -U none --no-map-gw --quiet --netns /run/user/1000/netns/netns-52b2cf2f-a904-7570-d531-3d0d0f243879
```

I also don't see anything in `coredumpctl list`.

**Comment by mtrmac** [comment]
> It’s far too attractive for me to suspect that “adding a `Printf` to a deferred function call” happened to turn one of the deferred calls into a non-deferred one.

So this _is_ true after all, but oh how hidden that was.

```diff
diff --git a/vendor/github.com/containers/image/v5/storage/storage_src.go b/vendor/github.com/containers/image/v5/storage/storage_src.go
index 967334c..6304743 100644
--- a/vendor/github.com/containers/image/v5/storage/storage_src.go
+++ b/vendor/github.com/containers/image/v5/storage/storage_src.go
@@ -161,9 +161,10 @@ func (s *storageImageSource) GetBlob(ctx context.Context, info types.BlobInfo, c
        if err != nil {
                return nil, 0, err
        }
+       rc2 := rc
        defer func() {
                fmt.Printf(">>> GetBlob defer #0, rc %#p\n", rc)
-               rc.Close()
+               rc2.Close()
        }()
 
        tmpFile, err := tmpdir.CreateBigFileTemp(s.systemContext, "")
```

`rc` is a named return value in this function, so `return ioutils.NewReadCloserWrapper` sets `rc`, and then that `defer` closes the temporary file instead of the input stream as intended.

`defer rc.Close()` evaluates `rc` immediately, and works; `defer func() { rc.Close() }()` evaluates `rc` at with the final return value, and doesn’t.

**Comment by debarshiray** [comment]
> Except, this time it's stuck inside `copyLayerFromStream()` within `blobPipelineDecryptionStep()`. This is puzzling because I was certain that `copyLayerFromStream()` wasn't being called before.

Ok, I was just misled by the progress bar overwriting the last line.

**Comment by mtrmac** [comment]
`skopeo … |& cat` to get rid of the progress bars.

**Comment by mtrmac** [comment]
> I can see one bug in there already, although that one should not make a difference. 

https://github.com/containers/image/pull/2394 for the record. Also removes the named return value gotcha.

**Comment by debarshiray** [comment]
> > It’s far too attractive for me to suspect that “adding a `Printf` to a deferred function call” happened to turn one of the deferred calls into a non-deferred one.
> 
> So this _is_ true after all, but oh how hidden that was.
> 
> ```diff
> diff --git a/vendor/github.com/containers/image/v5/storage/storage_src.go b/vendor/github.com/containers/image/v5/storage/storage_src.go
> index 967334c..6304743 100644
> --- a/vendor/github.com/containers/image/v5/storage/storage_src.go
> +++ b/vendor/github.com/containers/image/v5/storage/storage_src.go
> @@ -161,9 +161,10 @@ func (s *storageImageSource) GetBlob(ctx context.Context, info types.BlobInfo, c
>         if err != nil {
>                 return nil, 0, err
>         }
> +       rc2 := rc
>         defer func() {
>                 fmt.Printf(">>> GetBlob defer #0, rc %#p\n", rc)
> -               rc.Close()
> +               rc2.Close()
>         }()
>  
>         tmpFile, err := tmpdir.CreateBigFileTemp(s.systemContext, "")
> ```
> 
> `rc` is a named return value in this function, so `return ioutils.NewReadCloserWrapper` sets `rc`, and then that `defer` closes the temporary file instead of the input stream as intended.
> 
> `defer rc.Close()` evaluates `rc` immediately, and works; `defer func() { rc.Close() }()` evaluates `rc` with the final return value, and doesn’t.

Wow!  I did suspect that it might be related to the different evaluation of `rc` due to the change in the deferred call, but I didn't spot the named return value.  :)

Are those named return values really needed?  The way the function is written, they don't seem necessary at all.  At least we should use different names for them.

**Comment by mtrmac** [comment]
Also, I see
```
$ for i in $(seq 1 100); do (podman run --rm --detach       --name docker-registry       --publish 5000:5000       --rm       docker.io/library/registry:2 &&  skopeo copy --dest-creds=x       --dest-tls-verify=false       containers-storage:registry.fedoraproject.org/fedora:34       docker://localhost:5000/fedora:34 |& cat && podman stop -a && sleep 1) || break; done
```
passing, as an unprivileged user (with `su -` from a root’s session), with
```
# rpm -q passt podman skopeo
passt-0^20240426.gd03c4e2-1.fc40.x86_64
podman-5.0.2-1.fc40.x86_64
skopeo-1.15.0-1.fc40.x86_64
```

There must be some other variable involved if this is hanging frequently in other environments.

**Comment by debarshiray** [comment]
> > I can see one bug in there already, although that one should not make a difference.
> 
> [containers/image#2394](https://github.com/containers/image/pull/2394) for the record. Also removes the named return value gotcha.

Cool, I will take a look.  My GNOME Shell suddenly stopped accepting keyboard input, which seems to happen on Fedora 38.  So I had to reboot and missed your comment due to a stale page.

**Comment by debarshiray** [comment]
> There must be some other variable involved if this is hanging frequently in other environments.

I wonder if there's something about a virtual machine that triggers this.  I am using a Fedora 40 Workstation VM, and I expect the Toolbx CI running on Software Factory to be using VMs too.

**Comment by debarshiray** [comment]
> > A packet dump from the container's lo interface. That might be a bit fiddly to obtain if the container image doesn't include tcpdump, or dumpcap. It should be possible by using nsenter to enter just the container's netns and do that, though. There might be an easier podman specific way of doing so, @Luap99 ?
> 
> I really need to start writing some scripts around this given how common this is. The command would be somthing like `podman unshare nsenter --net=$(podman inspect --format {{.NetworkSettings.SandboxKey}} <container id/name>) <command inside netns...>` So if we only care about lo traffic then `tcpdump -i lo -w lo.pcap` as command in the container should do it.

I updated the previous URL with packets captured this way: https://rishi.fedorapeople.org/podman-22575-pasta.pcap

Note that it's 75M.  A lot bigger than the few kilobytes from the previous attempt.

I think another alternative would have been to create the Docker registry container with `--privileged` and then use `podman exec` to install `tcpdump(8)` and run it, but I don't know if that has any other consequences.

Anyway, after a very quick look ...

On the `skopeo(1)` client end, I see that it gets stuck at the first attempt at `makeRequestToResolvedURLOnce` and the debug output says:
```
DEBU[0002] No compression detected                      
DEBU[0002] Compressing blob on the fly
DEBU[0002] Uploading /v2/fedora/blobs/uploads/          
DEBU[0002] POST http://localhost:5000/v2/fedora/blobs/uploads/ 
DEBU[0002] PATCH http://localhost:5000/v2/fedora/blobs/uploads/bb6ad96b-cbba-4045-befe-bcf73c197cf3?_state=Frl_HMtQp2sVbHl5cQTfqBKekfeajGUpDb6tsbfx1mR7Ik5hbWUiOiJmZWRvc
```

On the Docker registry container side, I think I see the `PATCH` followed by an endless series of HTTP continuation packets.

**Comment by sbrivio-rh** [comment]
> ```
> DEBU[0002] PATCH http://localhost:5000/v2/fedora/blobs/uploads/bb6ad96b-cbba-4045-befe-bcf73c197cf3?_state=Frl_HMtQp2sVbHl5cQTfqBKekfeajGUpDb6tsbfx1mR7Ik5hbWUiOiJmZWRvc
> ```
> 
> On the Docker registry container side, I think I see the `PATCH` followed by an endless series of HTTP continuation packets.

This part looks fine in the capture, we have the three-way handshake in frames 21-23, a number of data transfers start, and the last one completes in frame 2268, getting a HTTP 202 (Accepted) response in frame 2269, then the connection closes.

If you look at it with Wireshark, you will see a number of seemingly missing packets ("TCP Previous segment not captured"), but I've already seen this happening and I think those packets were simply not captured by the kernel for whatever reason. In any case, pasta doesn't deal with TCP on this path, so it can't cause that kind of issue on its own.

What puzzles me is rather frame 4 (with Wireshark, `tcp.stream eq 0` as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8. Do you have a way to find out if this "bad request" is causing this issue? Does the client simply make a new attempt after that?

**Comment by debarshiray** [comment]
> What puzzles me is rather frame 4 (with Wireshark, tcp.stream eq 0 as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8.

I see.

> Do you have a way to find out if this "bad request" is causing this issue? Does the client simply make a new attempt after that?

I am trying to build Skopeo with HTTP [tracing](https://go.dev/blog/http-tracing) to get some more insight.

**Comment by debarshiray** [comment]
> > What puzzles me is rather frame 4 (with Wireshark, tcp.stream eq 0 as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8.
> 
> I see.

I wonder if I should run `pasta(1)` under Valgrind or such.  Is there an easy way to do that without hacking the `podman(1)` sources to insert the call to `valgrind(1)`?

**Comment by sbrivio-rh** [comment]
> > > What puzzles me is rather frame 4 (with Wireshark, tcp.stream eq 0 as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8.
> > 
> > 
> > I see.
> 
> I wonder if I should run `pasta(1)` under Valgrind or such. Is there an easy way to do that without hacking the `podman(1)` sources to insert the call to `valgrind(1)`?

You could use a shell script wrapper, but you can't easily run pasta under valgrind (yet) because it uses flags for `clone(2)` that are not supported by valgrind.

We have a set of [valgrind tests](https://passt.top/passt/tree/test/run?id=1ba76c9e8c14b21d8c2c7cb71abdaa85feb96605#n101) we routinely run with passt(1) (same binary, the command for virtual machines), and a special [target](https://passt.top/passt/tree/Makefile?id=1ba76c9e8c14b21d8c2c7cb71abdaa85feb96605#n129) in the Makefile as otherwise seccomp would terminate the process (the seccomp profile doesn't allow sbrk(2) and mmap(2), to ensure that passt and pasta don't dynamically allocate memory).

Those tests don't cover this particular path, though, because it's not used with virtual machines (we can't splice into a guest socket, of course). I checked this path with valgrind a couple of times in isolation (without the part that needs the clone() calls), but that's about it. This is definitely not ideal, and adding support for those clone() flags in valgrind is somewhere on my to-do list.

Give us a bit of time to look into those bytes, I'd say.

**Comment by debarshiray** [comment]
> What puzzles me is rather frame 4 (with Wireshark, `tcp.stream eq 0` as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8. Do you have a way to find out if this "bad request" is causing this issue? Does the client simply make a new attempt after that?

I wonder if that's related to the use of plain text HTTP instead of HTTPS.

The first HTTP request I have seen is:
```
GET /v2/ HTTP/1.1
Host: localhost:5000
User-Agent: skopeo/1.15.0
Docker-Distribution-Api-Version: registry/2.0
Accept-Encoding: gzip
```

... which prompts Go to throw a `tls.RecordHeaderError` with:
```
tls: first record does not look like a TLS handshake
```

I wish there was an easy way to trigger more detailed network logs from Skopeo instead of me doing things like hacking in custom [http.RounderTripper](https://pkg.go.dev/net/http#RoundTripper) implementations into the code.  :)

**Comment by debarshiray** [comment]
> > I wonder if I should run `pasta(1)` under Valgrind or such. Is there an easy way to do that without hacking the `podman(1)` sources to insert the call to `valgrind(1)`?
> 
> You could use a shell script wrapper, but you can't easily run pasta under valgrind (yet) because it uses flags for `clone(2)` that are not supported by valgrind.
> 
> We have a set of [valgrind tests](https://passt.top/passt/tree/test/run?id=1ba76c9e8c14b21d8c2c7cb71abdaa85feb96605#n101) we routinely run with passt(1) (same binary, the command for virtual machines), and a special [target](https://passt.top/passt/tree/Makefile?id=1ba76c9e8c14b21d8c2c7cb71abdaa85feb96605#n129) in the Makefile as otherwise seccomp would terminate the process (the seccomp profile doesn't allow sbrk(2) and mmap(2), to ensure that passt and pasta don't dynamically allocate memory).
> 
> Those tests don't cover this particular path, though, because it's not used with virtual machines (we can't splice into a guest socket, of course). I checked this path with valgrind a couple of times in isolation (without the part that needs the clone() calls), but that's about it. This is definitely not ideal, and adding support for those clone() flags in valgrind is somewhere on my to-do list.

Thanks for explaining all that!

> Give us a bit of time to look into those bytes, I'd say.

Okay, sure.

In the meantime, here's my latest instrumented build of Skopeo that shows the HTTP request and response headers, in case anybody finds it useful: https://koji.fedoraproject.org/koji/taskinfo?taskID=117353488

... and I would suggest trying a Fedora 40 virtual machine.  That's what the Toolbx CI uses, which triggered this bug in the first place.  I am able to reproduce at will using a F40 VM myself.  There are times when the `skopeo copy` does go through, but it's rare.

**Comment by dgibson** [comment]
> > ```
> > DEBU[0002] PATCH http://localhost:5000/v2/fedora/blobs/uploads/bb6ad96b-cbba-4045-befe-bcf73c197cf3?_state=Frl_HMtQp2sVbHl5cQTfqBKekfeajGUpDb6tsbfx1mR7Ik5hbWUiOiJmZWRvc
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On the Docker registry container side, I think I see the `PATCH` followed by an endless series of HTTP continuation packets.
> 
> This part looks fine in the capture, we have the three-way handshake in frames 21-23, a number of data transfers start, and the last one completes in frame 2268, getting a HTTP 202 (Accepted) response in frame 2269, then the connection closes.
> 
> If you look at it with Wireshark, you will see a number of seemingly missing packets ("TCP Previous segment not captured"), but I've already seen this happening and I think those packets were simply not captured by the kernel for whatever reason. In any case, pasta doesn't deal with TCP on this path, so it can't cause that kind of issue on its own.
> 
> What puzzles me is rather frame 4 (with Wireshark, `tcp.stream eq 0` as filter). TCP-wise, that's fine, but it looks like pasta is mangling some data there: we should have an HTTP request instead. The HTTP server answers 400 (Bad Request) in frame 8. Do you have a way to find out if this "bad request" is causing this issue? Does the client simply make a new attempt after that?

I'm wondering if this is an attempted tls/ssl handshake which the server is refusing because it's configured for plain http.  Wireshark is interpreting it as something differemt, but I don't know enough about tls at the bits & bytes level to say if it could just be making a bad guess at the protocol.  In any case I'm fairly skeptical that this is relevant to the overall problem, since this interaction seems to be over quickly and is followed by other more successful looking transactions.

The big thing I notice is that between frame 2227 and frame 2228 there's an approximately 70 second gap for no obvious reason.  There are a bunch of window updates just before that, so I wonder if something could be misinterpreting them as a zero window, even though they're clearly not.  That would have to be in the kernel, though, since for a splice connection like this pasta doesn't handle TCP at that level.

At this point what I kind of want is an strace of pasta, but that's a bit awkward to get, since it makes itself undumpable.

**Comment by sbrivio-rh** [comment]
I'm trying to reproduce this on a Fedora 40 virtual machine now, but meanwhile:

> At this point what I kind of want is an strace of pasta, but that's a bit awkward to get, since it makes itself undumpable.

...pasta can be straced without changes as root, too (`strace -f -p $(pgrep pasta)` or suchlike), or as user removing this line:

```
diff --git a/isolation.c b/isolation.c
index f394e93..780784f 100644
--- a/isolation.c
+++ b/isolation.c
@@ -377,8 +377,6 @@ void isolate_postfork(const struct ctx *c)
 {
        struct sock_fprog prog;
 
-       prctl(PR_SET_DUMPABLE, 0);
-
        if (c->mode == MODE_PASST) {
                prog.len = (unsigned short)ARRAY_SIZE(filter_passt);
                prog.filter = filter_passt;
```

Yes, at this point if you manage to strace it just before and during the hang, that would probably be really helpful.

**Comment by debarshiray** [comment]
I started the Docker registry container with:
```
$ podman run --name docker-registry --publish 5000:5000 --network pasta --rm docker.io/library/registry:2
...
```

Then attached `strace(1)` to `pasta(1)` as `root`:
```
# strace --attach 3597 --follow-forks --output podman-22575-pasta.strace --output-separately
strace: Process 3597 attached
```

Then tried a `skopeo copy`, which got stuck:
```
$ skopeo copy \
    --dest-tls-verify=false \
    containers-storage:registry.fedoraproject.org/fedora:40 \
    docker://localhost:5000/fedora:40
Getting image source signatures
Copying blob 77920da37914 done   |
```

Here's the strace log: https://rishi.fedorapeople.org/podman-22575-pasta.strace

It looks like there weren't any forks to follow after I attached `strace(1)` to `pasta(1)` because I only got one file.

**Comment by sbrivio-rh** [comment]
I haven't been able to reproduce this so far, I tried a Fedora 40 VM with:
```
$ rpm -q passt podman skopeo
passt-0^20240426.gd03c4e2-1.fc40.x86_64
podman-5.0.2-1.fc40.x86_64
skopeo-1.15.0-1.fc40.x86_64
$ uname -r
6.8.5-301.fc40.x86_64
```

and a Fedora Rawhide VM with:
```
# rpm -q passt podman skopeo
passt-0^20240426.gd03c4e2-1.fc41.x86_64
podman-5.0.2-1.fc41.x86_64
skopeo-1.15.0-1.fc41.x86_64
```

and two different kernel versions here, `6.9.0-0.rc0.20240315gite5eb28f6d1af.7.fc41.x86_64` and `6.9.0-0.rc6.20240503gitf03359bca01b.56.fc41.x86_64`. I also tried the loop from https://github.com/containers/podman/issues/22575#issuecomment-2093844750, no failures observed.

@debarshiray, how much memory and how many CPU threads did you give to your Fedora 40 VM? Available and total memory affect some TCP parameters -- so far I tried with 4 GiB and 1 GiB.

Other than the strace, there are two other bits of information (perhaps more convenient to fetch) that might help as well:

- pasta's tracing logs, which can be enabled by passing `--trace` and `--log-file` in pasta options, something like `podman run [...] --net=pasta:--trace,--log-file,/tmp/pasta_trace.log,--log-size,1000000000`. With the current version of Podman, that needs a small patch in Podman itself, because `--trace` can't be passed together with `--quiet` (option passed by Podman by default):
```
diff --git a/vendor/github.com/containers/common/libnetwork/pasta/pasta.go b/vendor/github.com/containers/common/libnetwork/pasta/pasta.go
index 0da7607f6..20d3d45d7 100644
--- a/vendor/github.com/containers/common/libnetwork/pasta/pasta.go
+++ b/vendor/github.com/containers/common/libnetwork/pasta/pasta.go
@@ -130,8 +130,7 @@ func Setup(opts *SetupOptions) error {
                cmdArgs = append(cmdArgs, "--dns", "none")
        }
 
-       // always pass --quiet to silence the info output from pasta
-       cmdArgs = append(cmdArgs, "--quiet", "--netns", opts.Netns)
+       cmdArgs = append(cmdArgs, "--netns", opts.Netns)
 
        logrus.Debugf("pasta arguments: %s", strings.Join(cmdArgs, " "))
 
```

- there's an optimisation in the TCP spliced path in pasta for bulk transfers, based on the `SO_RCVLOWAT` socket option, to avoid frequent socket wake-ups, and to get higher utilisation of the pipe using for splicing between sockets. This increases the complexity quite a bit, and I wonder if that's the culprit. Could you try to get rid of that in pasta:
```
diff --git a/tcp_splice.c b/tcp_splice.c
index 42b7be0..69487ec 100644
--- a/tcp_splice.c
+++ b/tcp_splice.c
@@ -592,17 +592,6 @@ eintr:
                        if (readlen >= (long)c->tcp.pipe_size * 10 / 100)
                                continue;
 
-                       if (conn->flags & lowat_set_flag &&
-                           readlen > (long)c->tcp.pipe_size / 10) {
-                               int lowat = c->tcp.pipe_size / 4;
-
-                               setsockopt(conn->s[fromside], SOL_SOCKET,
-                                          SO_RCVLOWAT, &lowat, sizeof(lowat));
-
-                               conn_flag(c, conn, lowat_set_flag);
-                               conn_flag(c, conn, lowat_act_flag);
-                       }
-
                        break;
                }
 
```
and see if you can still reproduce the issue? It might also be worth to get rid of the whole `if (readlen > 0 && readlen == written) {` block, actually, to reduce the complexity further (at the expense of throughput for bulk transfers).

**Comment by sbrivio-rh** [comment]
> Once it got stuck, I pressed `ctrl+c` to first terminate the `skopeo copy` and then the Docker registry container.
> 
> Here's the strace log: https://rishi.fedorapeople.org/podman-22575-pasta.strace

Hah, thanks, I missed you had already posted this while I was writing my previous comment. Could you please re-run strace with `-r` (`--relative-timestamps`) so that we can try to make sense of that ~ one minute gap in the traffic capture?

> It looks like there weren't any forks to follow after I attached `strace(1)` to `pasta(1)` because I only got one file.

Right, yes, that's pretty much expected, pasta clone()s periodically only if it needs to scan bound ports inside the container (for automatic port binding, which is not enabled by default with Podman), and occasionally to replenish pools of socket and pipes (the ones that need to be created inside the container), but if you have just a couple of connections, you won't see any cloned process.

**Comment by debarshiray** [comment]
I got:
```
$ uname --kernel-release
6.8.8-300.fc40.x86_64
$ rpm -q passt podman skopeo
passt-0^20240426.gd03c4e2-1.fc40.x86_64
podman-5.0.2-1.fc40.x86_64
skopeo-1.15.0-1.fc40.x86_64
```

My host has a AMD Ryzen 7 5800X CPU with 8 cores and 16 threads, and 32G of RAM.  It looks like my Fedora 40 Workstation virtual machine running on GNOME Boxes with QEMU/KVM has access to all the CPU cores and threads, and I gave it 4G of RAM.

See this screenshot from inside the VM, and I also verified from `/proc/cpuinfo` and `/proc/meminfo`:
![image](https://github.com/containers/podman/assets/3208107/b4d8b8ea-8c38-4b07-8645-2bb46e6b71c0)

**Comment by debarshiray** [comment]
> Could you please re-run strace with -r (--relative-timestamps) so that we can try to make sense of that ~ one minute gap in the traffic capture?

Yes, sure.  Updated with `--relative-timestamps`: https://rishi.fedorapeople.org/podman-22575-pasta.strace

**Comment by debarshiray** [comment]
> Other than the strace, there are two other bits of information (perhaps more convenient to fetch) that might help as well:
> 
>     * pasta's tracing logs, which can be enabled by passing `--trace` and `--log-file` in pasta options, something like `podman run [...] --net=pasta:--trace,--log-file,/tmp/pasta_trace.log,--log-size,1000000000`. With the current version of Podman, that needs a small patch in Podman itself, because `--trace` can't be passed together with `--quiet` (option passed by Podman by default):

Here's an instrumented Podman build: https://koji.fedoraproject.org/koji/taskinfo?taskID=117380638

> there's an optimisation in the TCP spliced path in pasta for bulk transfers, based on the SO_RCVLOWAT socket option, to avoid frequent socket wake-ups, and to get higher utilisation of the pipe using for splicing between sockets. This increases the complexity quite a bit, and I wonder if that's the culprit. Could you try to get rid of that in pasta:

Here's an instrumented Pasta build with the inner `SO_RCVLOWAT` socket option path removed: https://koji.fedoraproject.org/koji/taskinfo?taskID=117381587

Using those two, here's the `pasta --trace` log: https://rishi.fedorapeople.org/podman-22575-pasta.trace

**Comment by Luap99** [comment]
> Using those two, here's the `pasta --trace` log: https://rishi.fedorapeople.org/podman-22575-pasta.trace

This doesn't seem to be a public link, I get 403 Forbidden

**Comment by debarshiray** [comment]
> > Using those two, here's the `pasta --trace` log: https://rishi.fedorapeople.org/podman-22575-pasta.trace
> 
> This doesn't seem to be a public link, I get 403 Forbidden

Oops!  Fixed with a `chmod 644`.

Here's an instrumented Pasta build with the outer `if (readlen > 0 && readlen == written) {` block removed: https://koji.fedoraproject.org/koji/taskinfo?taskID=117382807

Here's the `pasta --trace` log with it: https://rishi.fedorapeople.org/podman-22575-pasta.2.trace

Looking closely at the Docker registry logs, I see these two extra lines as compared to [before](https://github.com/containers/podman/issues/22575#issuecomment-2090026483):
```
127.0.0.1 - - [07/May/2024:16:28:52 +0000] "PATCH /v2/fedora/blobs/uploads/098dd8dd-8f91-42d7-b4e3-281cad85b638?_state=wPOjVx-cQ2Cp28_n6NKbfGMYequei-3C3oDYtrzY_4V7Ik5hbWUiOiJmZWRvcmEiLCJVVUlEIjoiMDk4ZGQ4ZGQtOGY5MS00MmQ3LWI0ZTMtMjgxY2FkODViNjM4IiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDI0LTA1LTA3VDE2OjI4OjUyLjA2MjI4MjAyNloifQ%3D%3D HTTP/1.1" 202 0 "" "skopeo/1.15.0"
time="2024-05-07T16:29:34.420771963Z" level=info msg="response completed" go.version=go1.20.8 http.request.contenttype="application/octet-stream" http.request.host="localhost:5000" http.request.id=adb07195-a51c-474f-a680-66a635839110 http.request.method=PATCH http.request.remoteaddr="127.0.0.1:57252" http.request.uri="/v2/fedora/blobs/uploads/098dd8dd-8f91-42d7-b4e3-281cad85b638?_state=wPOjVx-cQ2Cp28_n6NKbfGMYequei-3C3oDYtrzY_4V7Ik5hbWUiOiJmZWRvcmEiLCJVVUlEIjoiMDk4ZGQ4ZGQtOGY5MS00MmQ3LWI0ZTMtMjgxY2FkODViNjM4IiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDI0LTA1LTA3VDE2OjI4OjUyLjA2MjI4MjAyNloifQ%3D%3D" http.request.useragent="skopeo/1.15.0" http.response.duration=42.270927123s http.response.status=202 http.response.written=0 
```

**Comment by debarshiray** [comment]
Those two lines are emitted by the Docker registry container *after* I hit `ctrl+c` to terminate the stuck `skopeo copy`.

So, the Docker registry container claims to have completed responding to the `PATCH` request after the `skopeo(1)` client terminates.  However, the instrumented `skopeo(1)` only shows the `PATCH` request, but never any response.

**Comment by sbrivio-rh** [comment]
Thanks for the additional traces, I'm looking into them now. And:

> It looks like my Fedora 40 Workstation virtual machine running on GNOME Boxes with QEMU/KVM has access to all the CPU cores and threads

...it looks like this was the key to reproduce it. I assigned CPU 12 threads (instead of one) to my Fedora 40 virtual machine and I can finally reproduce the hang as well! For some reason, it happens only once per boot. If I `podman stop` the container and then restart it, the next attempt will work for me. If I reboot, it hangs on the first attempt (three out of three attempts).

At least on my setup, I see this:

```
     0.000059 epoll_wait(4, [{events=EPOLLIN, data={u32=54018, u64=8589988610}}], 8, 1000) = 1
     0.000053 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 73728
     0.000059 splice(150, NULL, 115, NULL, 73728, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 73728
     0.000063 epoll_wait(4, [{events=EPOLLIN, data={u32=54018, u64=8589988610}}], 8, 1000) = 1
     0.000055 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 73728
     0.000046 splice(150, NULL, 115, NULL, 73728, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 65320
     0.000043 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 65536
     0.000048 splice(150, NULL, 115, NULL, 73944, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
     0.000055 epoll_wait(4, [{events=EPOLLIN, data={u32=54018, u64=8589988610}}], 8, 1000) = 1
     0.000049 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 16770
     0.000066 splice(150, NULL, 115, NULL, 16770, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
```

everything looks normal until this point. Twice in a row, the receiver (socket 115) is not ready to read what we're splicing (pipe ends: 151 to 150)  from the originating socket (211). Then we wait until:

```
     0.000061 epoll_wait(4, [{events=EPOLLOUT, data={u32=29442, u64=12884931330}}], 8, 1000) = 1
     0.005003 epoll_ctl(4, EPOLL_CTL_MOD, 211, {events=EPOLLIN|EPOLLRDHUP, data={u32=54018, u64=8589988610}}) = 0
     0.000089 epoll_ctl(4, EPOLL_CTL_MOD, 115, {events=EPOLLIN|EPOLLRDHUP, data={u32=29442, u64=12884931330}}) = 0
     0.000081 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
     0.000073 splice(150, NULL, 115, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 1048576
     0.000087 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
     0.000045 splice(150, NULL, 115, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = 520415
     0.000060 splice(211, NULL, 151, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
     0.000044 splice(150, NULL, 115, NULL, 1048576, SPLICE_F_MOVE|SPLICE_F_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
     0.000044 epoll_wait(4, [], 8, 1000) = 0
     1.001310 epoll_wait(4, [], 8, 1000) = 0
```

we get `EPOLLOUT` on the receiver side, meaning we can finally write more. But right after that, we can't read anything from the sender (socket 211), which is rather unexpected. Full strace from just before issuing the skopeo command to the hang at https://paste.ec/paste/RNcEEQLw#7hmpI+MPxIYkQhQUvKsfpzSR4O2K7wndNoX31m2Ah0B

**Comment by sbrivio-rh** [comment]
I just posted a patch at https://archives.passt.top/passt-dev/20240508090338.2735208-1-sbrivio@redhat.com/ that fixes the issue for me.

Short summary: we fill up the buffer of the receiving socket (here, container-side), so we get `EAGAIN`. On the sending socket (here, host-side), we just got `EAGAIN` as well, because the sender has nothing more to send. We conclude that we have nothing to transfer. But that's not the case: there are still bytes on the pipe that need to be written to the receiving socket, so we should subscribe (again) to `EPOLLOUT` events there. Full details in the commit message.

@debarshiray, it would be great if you could test this: `curl https://archives.passt.top/passt-dev/20240508090338.2735208-1-sbrivio@redhat.com/raw | git am` if you have a git tree, but it's anyway a one-line patch so it can be applied manually too. Thanks!

**Comment by debarshiray** [comment]
> I just posted a patch at https://archives.passt.top/passt-dev/20240508090338.2735208-1-sbrivio@redhat.com/ that fixes the issue for me.
> 
> [...]
> 
> @debarshiray, it would be great if you could test this: `curl https://archives.passt.top/passt-dev/20240508090338.2735208-1-sbrivio@redhat.com/raw | git am` if you have a git tree, but it's anyway a one-line patch so it can be applied manually too. Thanks!

Yes, the patch appears to fix this problem for me too!

I tried three times on my Fedora 40 VM, rebooted and tried three times again.  It didn't get stuck.

**Comment by sbrivio-rh** [comment]
This is now fixed in passt version [2024_05_10.7288448](https://archives.passt.top/passt-user/20240510225714.6aa8e6c0@elisabeth/) and in the corresponding updates for [Fedora 39](https://bodhi.fedoraproject.org/updates/FEDORA-2024-97286d7a34), [Fedora 40](https://bodhi.fedoraproject.org/updates/FEDORA-2024-2ceb4e8047), and [Fedora Rawhide](https://bodhi.fedoraproject.org/updates/FEDORA-2024-0133ac6761).

**Comment by debarshiray** [comment]
Closing.

**Comment by debarshiray** [comment]
Thanks for all the work to fix this.  Much appreciated.
